api/v1alpha1/linodevpc_webhook.go

```diff
@@ -0,0 +1,264 @@
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha1
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/netip"
+	"regexp"
+	"slices"
+	"strings"
+
+	"go4.org/netipx"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+
+	. "github.com/linode/cluster-api-provider-linode/clients"
+)
+
+var (
+	// The capability string indicating a region supports VPCs: [VPC Availability]
+	//
+	// [VPC Availability]: https://www.linode.com/docs/products/networking/vpc/#availability
+	LinodeVPCCapability = "VPCs"
+
+	// The IPv4 ranges that are excluded from VPC Subnets: [Valid IPv4 Ranges for a Subnet]
+	//
+	// [Valid IPv4 Ranges for a Subnet]: https://www.linode.com/docs/products/networking/vpc/guides/subnets/#valid-ipv4-ranges
+	LinodeVPCSubnetReserved = mustParseIPSet("192.168.128.0/17")
+
+	// IPv4 private address space as defined in [RFC 1918].
+	//
+	// [RFC 1918]: https://datatracker.ietf.org/doc/html/rfc1918
+	privateIPv4 = mustParseIPSet("10.0.0.0/8", "172.16.0.0/12", "192.168.0.0/16")
+)
+
+// mustParseIPSet parses the given IP CIDRs as a [go4.org/netipx.IPSet]. It is intended for use with hard-coded strings.
+//
+//nolint:errcheck //^
+func mustParseIPSet(cidrs ...string) *netipx.IPSet {
+	var (
+		builder netipx.IPSetBuilder
+		set     *netipx.IPSet
+	)
+	for _, cidr := range cidrs {
+		prefix, _ := netip.ParsePrefix(cidr)
+		builder.AddPrefix(prefix)
+	}
+	set, _ = builder.IPSet()
+	return set
+}
+
+// log is for logging in this package.
+var linodevpclog = logf.Log.WithName("linodevpc-resource")
+
+// SetupWebhookWithManager will setup the manager to manage the webhooks
+func (r *LinodeVPC) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// TODO(user): change verbs to "verbs=create;update;delete" if you want to enable updation and deletion validation.
+//+kubebuilder:webhook:path=/validate-infrastructure-cluster-x-k8s-io-v1alpha1-linodevpc,mutating=false,failurePolicy=fail,sideEffects=None,groups=infrastructure.cluster.x-k8s.io,resources=linodevpcs,verbs=create,versions=v1alpha1,name=vlinodevpc.kb.io,admissionReviewVersions=v1
+
+var _ webhook.Validator = &LinodeVPC{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *LinodeVPC) ValidateCreate() (admission.Warnings, error) {
+	linodevpclog.Info("validate create", "name", r.Name)
+
+	ctx, cancel := context.WithTimeout(context.Background(), defaultWebhookTimeout)
```

Comment:
can we move this code into the validateLinodeVPC function since we cannot test this code here?

Reply:
I believe the convention around contexts is to create the context as early as possible as pass it around?

Reply:
that assumes multiple things are using the context though, in this case, `ValidateCreate` is just a wrapper around the `validateLinodeVPC` function

Reply:
1. That seems like an internal implementation detail. I think we should keep the webhook framework simple, yet flexible enough to accommodate for future changes if we do decide to have a separate context for different operations, etc.

2. We do make use of passing different contexts to `validateLinodeVPC` in the unit tests.

Reply:
any code directly in the validate{Create,Update,Delete} functions is untestable. keeping as much logic as possible out of these functions and in the functions with the linode clients will keep code more easy to write tests for

Reply:
ah, I now understand what you mean by the different contexts in the unit tests, this sounds good